### PR TITLE
[Tests] Fixing leaky tests

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
### Motivation and Context

Trying to lower the memory usage of our test suite by identifying leaky tests

### Description

We have a big problem in our application that running all client test suites requires an absurd amount of memory. One of the reasons for this are leaky tests. What are memory leaks in JS and what causes them?

#### **_What are memory leaks?_**

**A very good explanation that you should definitely read to understand the problem: https://auth0.com/blog/four-types-of-leaks-in-your-javascript-code-and-how-to-get-rid-of-them/**

In essence:

- JS is a **garbage collected language**
- Modern garbage collectors improve on this algorithm in different ways, but the essence is the same: **reachable pieces of memory are marked as such and the rest is considered garbage.**
- Unwanted references are references to pieces of memory that the developer knows he or she won't be needing anymore but that for some reason are kept inside the tree of an active root. **In the context of JavaScript, unwanted references are variables kept somewhere in the code that will not be used anymore and point to a piece of memory that could otherwise be freed.** Some would argue these are developer mistakes.

I found a way to manually request the garbage collection after jest runs a test. 
If we look into the jest source code we find the following condition: https://github.com/facebook/jest/blob/40b8e1e157c9981dda5a68d73fff647e80fc9f5c/packages/jest-runner/src/runTest.ts#L291
image

This means by setting both the `--expose-gc` in node and `--logHeapUsage` in jest, a **minor** garbage collection will be **requested** by jest after each test. Requested means that the runtime will schedule one to occur soon. This should fix our leaky test problem and dramatically lower the memory usage on bamboo etc. There will probably be some performance cost but imo it will be worth it.

I therefore changed the testing commands to :

` "test": "node --expose-gc ./node_modules/.bin/jest  --logHeapUsage --config ./src/test/javascript/jest.config.js --env=jsdom -w=4"`

 `"test:ci": "node --expose-gc ./node_modules/.bin/jest  --logHeapUsage --config ./src/test/javascript/jest.config.ci.js --env=jsdom -w=1 --coverage"`

` "test:coverage": "node --expose-gc ./node_modules/.bin/jest  --logHeapUsage --config ./src/test/javascript/jest.config.js --env=jsdom -w=4 --coverage"`

The --expose-gc option will make the function gc() available on the global object, which the custom command needs in order to request the garbage collection.

**I can confirm in tests that this seem to fix our memory leak problem as the heap does not go over 700mb. For example see the test log here: https://pastebin.com/raw/JR6HZ68N**

The crucial commit is this one: dbb08f1a22884587284ab2921065541cba4774d7

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -75,6 +75,7 @@
         "split.js": "1.6.2",
         "ts-cacheable": "1.0.2",
         "tslib": "2.0.3",
+        "weak-napi": "^2.0.2",
```
Comment (krusche) [inline package.json:78]:
```suggestion
        "weak-napi": "2.0.2",
```

Comment (krusche) [reply package.json:78]: and please put it into devDependencies

Comment (stefanwaldhauser) [reply package.json]: Done in 089e7020eedb5ed6cc98a7425f377195048d72ee

**Review by krusche — APPROVED** [review]
Looks good. Thanks for this nice contribution 👍

**Pull request merged**